app/models/pipeline_run.rb

```diff
@@ -411,6 +424,60 @@ def db_load_ercc_counts
     update(total_ercc_reads: total_ercc_reads)
   end
 
+  def db_load_insert_size_metrics
+    insert_size_metrics_s3_path = "#{host_filter_output_s3_path}/#{INSERT_SIZE_METRICS_OUTPUT_NAME}"
+    _stdout, _stderr, status = Open3.capture3("aws", "s3", "ls", insert_size_metrics_s3_path)
+    return unless status.exitstatus.zero?
+    insert_size_metrics_lines = Syscall.pipe_with_output(["aws", "s3", "cp", insert_size_metrics_s3_path, "-"])
+    tsv_lines = []
+    tsv_header_line = -1
+    insert_size_metrics_lines.split(/\r?\n/).each_with_index do |line, index|
+      if line.start_with?("## METRICS CLASS")
+        tsv_header_line = index
+      elsif tsv_header_line > 0 && index - tsv_header_line <= 2
+        tsv_lines << CSV.parse_line(line, col_sep: "\t")
+      elsif tsv_lines.length >= 2
+        break
+      end
+    end
+    if tsv_lines.length != 2
+      Rails.logger.error("Pipleine run ##{id} has an insert size metrics file but metrics could not be found")
+      return
+    end
+    insert_size_metrics = {}
+    tsv_lines[0].zip(tsv_lines[1]).each do |row|
+      insert_size_metrics[row[0]] = row[1]
+    end
+
+    extract_int = lambda do |metrics, metric_name|
```

Comment:
To be more Ruby like, I would suggest not using a lambda here. That makes the code harder to use since you are forced to use `call`.
You don't have any closures from the parent method, so I would just declare a private method in the same class.

However, if I understood it right, you are just trying to prevent errors when metric_name is not present, correct?
In this case, you don't actually need this rescue block at all:

```
   # String to_i returns 0
   "abc".to_i
    # 0

    # nil to_i returns 0
    nil.to_i
    # 0

    # non-existent key returns nil, therefore it will return 0
    a = {"b": 1}
    a["c"].to_i
    # 0
```

Which means this rescue block will never get executed, so the result will never be nil. Eventually it will be 0.

Reply:
Since this is an optional value and for our reference values it is 0 I want to distinguish between no value and 0. For our artificial samples the value actually is 0. If these keys weren't found in the file for whatever reason I don't want it to show up as 0. I will fix my check and use a private method.